File: lua/hello_grpc.lua

pactLua = require("pactLua")
local ffi = require("ffi")
JSON = require("JSON")
local version = ffi.string(lib.pactffi_version())
local appName = ffi.string(lib.pactffi_version())

-- Setup Loggers
lib.pactffi_logger_init()
lib.pactffi_logger_attach_sink('stdout', 3)
lib.pactffi_logger_attach_sink('stderr', 2)
lib.pactffi_logger_apply()
lib.pactffi_log_message(appName, 'INFO', 'hello from ffi version: ' .. ffi.string(lib.pactffi_version()))


dir = os.getenv("PWD") or io.popen("cd"):read()
pact_contents =  {
    ["pact:proto"] = dir.."/../proto/area_calculator.proto",
    ["pact:proto-service"] = "Calculator/calculateOne",
    ["pact:content-type"] = "application/protobuf",
    request = {
      rectangle = {
        length = "matching(number, 3)",
        width = "matching(number, 4)"
      }
    },
    response = {
      value = { "matching(number, 12)" }
    }
  }
local contents = JSON:encode(pact_contents)

lib.pactffi_log_message(appName, 'INFO', "Loading pact contents: "..contents)

-- Setup pact for testing
pact = lib.pactffi_new_pact('grpc-consumer-lua', 'area-calculator-lua')
lib.pactffi_with_pact_metadata(pact, 'pact-lua', 'ffi', version)
message_pact = lib.pactffi_new_sync_message_interaction(pact, 'A gRPC calculateOne request')
lib.pactffi_with_specification(pact, 5)


-- Start mock server
lib.pactffi_using_plugin(pact, 'protobuf', '0.3.14')
lib.pactffi_interaction_contents(message_pact, 0, 'application/grpc',contents)
mock_server_port = lib.pactffi_create_mock_server_for_transport(pact , '0.0.0.0',0,'grpc',ffi.cast("void *", 0))
lib.pactffi_log_message(appName, 'INFO', "Mock server started: "..mock_server_port)

-- Make our client call
--  expected_response = 12.0
--  response = get_rectangle_area(f"localhost:{mock_server_port}")
--  print(f"Client response: {response}")
--  print(f"Client response - matched expected: {response == expected_response}")

result = lib.pactffi_mock_server_matched(mock_server_port)
if result then
    lib.pactffi_log_message(appName, 'INFO', "Pact - Got matching client requests: "..result)
    PACT_FILE_DIR='./pacts'
    lib.pactffi_log_message(appName, 'INFO', "Writing pact file to "..PACT_FILE_DIR)
    res_write_pact = lib.pactffi_write_pact_file(mock_server_port, PACT_FILE_DIR, false)
    lib.pactffi_log_message(appName, 'INFO', "Pact file writing results "..res_write_pact)
else
    print('pactffi_mock_server_matched did not match')
    lib.pactffi_log_message(appName, 'INFO', "pactffi_mock_server_matched did not match")
    mismatchers = lib.pactffi_mock_server_mismatches(mock_server_port)
    if mismatchers then
        lib.pactffi_log_message(appName, 'INFO', "pactffi_mock_server_matched did not match"..ffi.string(mismatchers))
    end
end
--  Cleanup
lib.pactffi_cleanup_mock_server(mock_server_port)
lib.pactffi_cleanup_plugins(pact)

File: php/hello_grpc.php
<?php
echo php_uname();
echo PHP_OS;

$library = php_uname('s') == 'Darwin' ? 'libpact_ffi.dylib' : (strtoupper(substr(PHP_OS, 0, 3)) === 'WIN' ? 'pact_ffi.dll' : 'libpact_ffi.so');
$code = file_get_contents(__DIR__ . '/../pact.h');
$ffi = FFI::cdef($code, __DIR__ . '/../' . $library);

// Setup Loggers
$ffi->pactffi_logger_init();
// $ffi->pactffi_logger_attach_sink('file ./logs/log-info.txt',5);
// $ffi->pactffi_logger_attach_sink('file ./logs/log-error.txt',5);
$ffi->pactffi_logger_attach_sink('stdout', 3);
$ffi->pactffi_logger_attach_sink('stderr', 2);
$ffi->pactffi_logger_apply();
$ffi->pactffi_log_message('pact-php-ffi', 'INFO', 'hello from pact php ffi, using Pact FFI Version: ' . $ffi->pactffi_version());

// Setup pact for testing
$pact = $ffi->pactffi_new_pact('grpc-consumer-php', 'area-calculator-provider');
$ffi->pactffi_with_pact_metadata($pact, 'pact-php','ffi',$ffi->pactffi_version());
$message_pact = $ffi->pactffi_new_sync_message_interaction($pact, 'A gRPC calculateOne request');
$ffi->pactffi_with_specification($pact, $ffi->PactSpecification_V4);

// Setup contents

$proto_file_path = __DIR__ . '/../proto/area_calculator.proto';

$contents = '{
    "pact:proto":  "'. $proto_file_path . '",
    "pact:proto-service": "Calculator/calculateOne",
     "pact:content-type": "application/protobuf",
     "request": {
       "rectangle": {
         "length": "matching(number, 3)",
         "width": "matching(number, 4)"
       }
     },
     "response": {
       "value": ["matching(number, 12)"]
     }
   }';

// Start mock server

$ffi->pactffi_using_plugin($pact, 'protobuf', '0.3.14');
$ffi->pactffi_interaction_contents($message_pact, 0, 'application/grpc', $contents);
$port = $ffi->pactffi_create_mock_server_for_transport($pact , '0.0.0.0',0,'grpc', null);

echo sprintf("Mock server port=%d\n", $port);

// This is where we would call our client, gRPC in this example plugin demo
// PHP gRPC is client only, so would need to use a provider from the following
// https://github.com/you54f/pact-plugins/tree/main/examples/gRPC/area_calculator
// TODO build out an area calculator PHP example
// https://grpc.io/docs/languages/php/

// Check if requests match - Note this _should_ fail, but the pactffi_mock_server_matched is returning true
// even if no requests were made.

if ($ffi->pactffi_mock_server_matched($port)) {
    echo getenv('MATCHING') ? "Mock server matched all requests, Yay!" : "Mock server matched all requests, That Is Not Good (tm)";
    echo "\n";

    $ffi->pactffi_write_pact_file($port, __DIR__ . '/../pacts', false);
    // $ffi->pactffi_write_message_pact_file($messagePact, __DIR__ . '/../pacts', false);
} else {
    echo getenv('MATCHING') ? "We got some mismatches, Boo!" : "We got some mismatches, as expected.";
    echo "\n";
    echo sprintf("Mismatches: %s\n", print_r(json_decode(FFI::string($ffi->pactffi_mock_server_mismatches($port)), true), true));
}

$ffi->pactffi_cleanup_mock_server($port);
$ffi->pactffi_cleanup_plugins($pact);
